File: backend/services/index_service.py
# backend/services/index_service.py
"""
此模块提供指数数据相关的服务功能。
包括获取指数列表、指数详情和K线数据的服务方法。
Authors: hovi.hyw & AI
Date: 2025-03-12
"""

from datetime import date, timedelta
from sqlalchemy.orm import Session

from backend.database.queries import get_index_list, get_index_kline_data, get_index_info


class IndexService:
    """
    指数数据服务类。
    提供获取指数列表、指数详情和K线数据的服务方法。

    Methods:
        get_index_list: 获取指数列表
        get_index_info: 获取指数详情
        get_index_kline: 获取指数K线数据

    Examples:
        >>> from sqlalchemy.orm import Session
        >>> service = IndexService()
        >>> indices = service.get_index_list(db, page=1, page_size=10)
    """

    def get_index_list(self, db: Session, page_size: int = 20, cursor: str | None = None, search: str | None = None, page: int | None = None):
        """
        获取指数列表。

        Args:
            db (Session): 数据库会话
            page_size (int): 每页数量，默认为20
            cursor (str, optional): 分页游标，用于获取下一页数据
            search (str, optional): 搜索关键字
            page (int, optional): 页码，从1开始，与cursor互斥，优先使用page

        Returns:
            dict: 包含指数列表和分页信息的字典
        """
        # 安全处理search参数
        search_str = ""
        if search is not None:
            # 直接尝试转换为字符串，不管是什么类型
            try:
                search_str = str(search)
                # 如果是空字符串或只包含空格，则设为空字符串
                if not search_str.strip():
                    search_str = ""
            except Exception:
                # 如果转换失败，使用空字符串
                search_str = ""
        
        return get_index_list(db, page_size, cursor, search_str, page)

    def get_index_info(self, db: Session, symbol: str):
        """
        获取指数详情。

        Args:
            db (Session): 数据库会话
            symbol (str): 指数代码

        Returns:
            dict: 指数详情信息

        Raises:
            ValueError: 如果指数不存在
        """
        index_info = get_index_info(db, symbol)
        if not index_info:
            raise ValueError(f"Index with symbol {symbol} not found")
        return index_info

    def get_index_kline(self, db: Session, symbol: str, start_date: date | None = None, end_date: date | None = None):
        """
        获取指数K线数据。

        Args:
            db (Session): 数据库会话
            symbol (str): 指数代码
            start_date (date, optional): 开始日期，默认为None（获取所有数据）
            end_date (date, optional): 结束日期，默认为None（获取所有数据）

        Returns:
            dict: 包含指数代码、名称和K线数据的字典

        Raises:
            ValueError: 如果未找到数据
        """
        # 调用queries.py中的函数获取K线数据
        kline_data = get_index_kline_data(db, symbol, start_date, end_date)

        if not kline_data:
            raise ValueError(f"No data found for index {symbol} in the specified date range")

        # 获取指数名称
        index_info = get_index_info(db, symbol)
        name = index_info.get('name') if index_info else None

        return {
            "symbol": symbol,
            "name": name,
            "data": kline_data,
        }

File: frontend/src/services/indexService.js
// frontend/src/services/indexService.js
/**
 * 此模块提供指数数据相关的服务。
 * 封装了获取指数列表、详情、K线数据和真实涨跌数据的方法。
 * Authors: hovi.hyw & AI
 * Date: 2025-03-12
 * 更新: 2025-03-16 - 添加获取指数真实涨跌数据的方法
 */

import api from './api';

/**
 * 获取指数列表
 * @param {Object} params - 请求参数
 * @param {string} params.cursor - 分页游标，用于获取下一页数据
 * @param {number} params.page - 页码，从1开始
 * @param {number} params.page_size - 每页数量，默认20
 * @param {string} params.search - 搜索关键字
 * @returns {Promise<Object>} 指数列表数据
 */
export const getIndexList = async (params = {}) => {
  const { cursor, page, page_size = 20, search = '' } = params;
  const queryParams = {
    page_size,
    ...(cursor ? { cursor } : {}),
    ...(page ? { page } : {}),
    ...(search ? { search } : {})
  };

  try {
    const response = await api.get('/indices', { params: queryParams });
    return response;
  } catch (error) {
    console.error('获取指数列表失败:', error);
    throw error;
  }
};

/**
 * 获取指数详情
 * @param {string} symbol - 指数代码
 * @returns {Promise<Object>} 指数详情数据
 */
export const getIndexInfo = async (symbol) => {
  try {
    const response = await api.get(`/indices/${symbol}`);
    return response;
  } catch (error) {
    console.error(`获取指数${symbol}详情失败:`, error);
    throw error;
  }
};

/**
 * 获取指数K线数据
 * @param {string} symbol - 指数代码
 * @param {Object} params - 请求参数
 * @param {string} params.start_date - 开始日期 (YYYY-MM-DD)
 * @param {string} params.end_date - 结束日期 (YYYY-MM-DD)
 * @returns {Promise<Object>} 指数K线数据
 */
export const getIndexKline = async (symbol, params = {}) => {
  try {
    const response = await api.get(`/indices/${symbol}/kline`, { params });
    return response;
  } catch (error) {
    console.error(`获取指数${symbol}K线数据失败:`, error);
    throw error;
  }
};

/**
 * 获取指数涨跌幅数据
 * @param {string} symbol - 指数代码
 * @param {Object} params - 请求参数
 * @param {string} params.start_date - 开始日期 (YYYY-MM-DD)
 * @param {string} params.end_date - 结束日期 (YYYY-MM-DD)
 * @returns {Promise<Object>} 指数涨跌幅数据
 */
export const getIndexChangeRate = async (symbol, params = {}) => {
  try {
    // 使用正确的指数API路径格式
    const response = await api.get(`/indices/${symbol}/real-change`, { params });
    return response;
  } catch (error) {
    console.error(`获取指数${symbol}真实涨跌数据失败:`, error);
    throw error;
  }
};